### PR TITLE
[BE] Request logging

### PR DESCRIPTION
**Pull request by qroll**
It'd be good to introduce structured logging so that it's easier to trace logs in production. Each log should minimally identify the originating request (uuid), have a timestamp (Cloud Run seems to append that already), and have a format that is easy to parse/search (JSON?).

This branch is kind of a POC?

- Wanted some thoughts on using AsyncLocalStorage as a way to track the current request context across nested calls without having to explicitly pass down the request object.
- From what I understand, [Cloud Logging will be able to parse serialised JSON](https://cloud.google.com/run/docs/logging#using-json) and extract special fields. I'm using console.dir right now but since it does not output valid JSON, not the best choice here.

### CODE REVIEW

File: backend/src/helpers/logger.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+export const LoggerContextMiddleware: RequestHandler = (req, res, next) => {
+	const context = {
+		id: crypto.randomUUID(),
+		req,
+	};
+
+	requestContextLocalStorage.run(context, () => {
+		Logger.log("Request start", {
+			body: req.body,
```
Comment (stonefruit) [inline backend/src/helpers/logger.ts:59]: Need to redact sensitive info over here

**Comment by stonefruit** [comment]
It might be easier to replace `console.log = Logger.log` when importing the logger file, then logger won't need to be imported everywhere.

Might also want to include timestamp instead of relying on gcloud's one as it might be useful for local development.

FYI in case it needed, `NODE_ENV` values `gcloud-development` and `production` are used for current gcloud applications

File: backend/src/helpers/logger.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+};
+
+// set up a store to persist data through async operations, making them accessible from nested calls
+export const requestContextLocalStorage =
+	new AsyncLocalStorage<RequestContext>();
+
+export class Logger {
+	static log(message: string, ...data: any[]) {
+		const context = requestContextLocalStorage.getStore();
+		console.dir(
```
Comment (stonefruit) [inline backend/src/helpers/logger.ts:17]: Actually should redact here instead just before anything actually gets logged.

**Comment by qroll** [comment]
It might be safer not to monkey patch `console.log` in case other libs are relying on it, or if we ever need to call the original function. Alternatively, webpack's [ProvidePlugin](https://webpack.js.org/plugins/provide-plugin/) lets you load the module without explicitly importing it.

**Review by ryhuz — APPROVED** [review]

**Pull request merged**